File: sdk/examples/python/approval_gate.py
from __future__ import annotations

import anyio

from roder_sdk import InMemoryTransport, RoderAgent


async def main() -> None:
    async def handler(request: dict) -> dict:
        if request["method"] == "thread/start":
            return {"jsonrpc": "2.0", "id": request["id"], "result": {"thread": {"id": "thread-approval"}}}
        if request["method"] == "turn/start":
            return {"jsonrpc": "2.0", "id": request["id"], "result": {"turn": {"id": "turn-approval"}}}
        return {"jsonrpc": "2.0", "id": request["id"], "result": {"ok": True}}

    async def on_tool_approval(request: dict) -> dict:
        return {"approved": request.get("toolName") == "fs/readFile", "message": "example policy"}

    agent = await RoderAgent.create(
        transport=InMemoryTransport(handler),
        approvals={"on_tool_approval": on_tool_approval},
    )
    await agent.send("Read a file only if policy allows it")
    await agent.close()


anyio.run(main)

File: sdk/examples/typescript/approval-gate.ts
import { InMemoryTransport, RoderAgent } from "../../typescript/src/index.js";

const transport = new InMemoryTransport((request) => ({
  jsonrpc: "2.0",
  id: request.id,
  result:
    request.method === "thread/start"
      ? { thread: { id: "thread-approval" } }
      : request.method === "turn/start"
        ? { turn: { id: "turn-approval" } }
        : { ok: true },
}));

const agent = await RoderAgent.create({
  transport,
  approvals: {
    onToolApproval(request) {
      const toolName = (request as { toolName?: string }).toolName;
      return { approved: toolName === "fs/readFile" };
    },
  },
});

await agent.send("Read a file only if policy allows it");
transport.emit({
  jsonrpc: "2.0",
  method: "thread/approvalRequested",
  params: { approvalId: "approval-example", toolName: "fs/readFile" },
});
await agent.close();
